### PR TITLE
feat(repo): add coded-term autocomplete to SOAP notes

### LINKED ISSUE #2487: PIMS: coded-term autocomplete on SOAP notes (vocabulary references per S/O/A/P section)
## Summary

Add coded-term autocomplete to SOAP notes: per-section pickers in the appointment workspace that search the Yosemite clinical vocabulary and attach exact vocabulary references (YC codes) to the note.

## Problem / motivation

The terminology programme shipped the vocabulary (11,742 terms), multilingual synonym search, the VeNom hierarchy, and cross-vocabulary mappings — but SOAP notes remained free text. On dev: 62 SOAP notes, 0 with any coded reference; `GET /v1/codes/terms/suggest` has no UI consumer; the `SoapNote.diagnoses` JSON channel round-trips through the FHIR mapper but nothing writes it. A vet typing "vomiting" produces a string — nothing links it to YC-005423, nothing rolls up in analysis, nothing exports as SNOMED/VeNom.

## Proposed solution

- A `SoapCodedTermPicker` chip-row under each native S/O/A/P rich-text section: type ≥2 chars → ranked suggestions (display + multilingual synonyms) → pick → chip with label + code, removable.
- Domain scoping per section: Subjective→PresentingComplaint, Assessment→Diagnosis, Plan→Procedure, Objective→all domains.
- Persist picks per section in the existing `diagnoses` channel (`soap-note-diagnoses` composition extension) as a shared, sanitized `SoapCodedProblems` shape; rehydrate on reload; list in the signed-note history read-out.
- FHIR consumers see the codes in the SOAP Composition's `soap-note-diagnoses` extension (already emitted by the shared mapper).

## Scope

Web PIMS appointment workspace, native four-section SOAP layout.

### Out of scope

- Species filtering (vocabulary species are SA/LA/EQUINE… practice buckets; no companion→bucket mapping exists yet).
- Coded terms in the merged clinical-packet PDF / discharge summary.
- Inline @-style mentions inside the rich-text editors (chip rows first).
- Mobile SOAP surfaces; custom-template SOAP notes.

## Current state / where it plugs in

`SoapStep`/`NativeSoapFields` (workspace SOAP step), `workspaceClinicalService` save/rehydrate, `packages/types` clinical-artifact mapper shapes, backend `/v1/codes/terms/suggest` (live on dev, multilingual).

## Acceptance criteria

- Typing in a section picker returns ranked suggestions from the deployed vocabulary, filtered by that section's domain; synonym hits show which synonym matched.
- Picked chips persist across save + reload and appear per-section in the signed-note history.
- Removing every chip from a persisted draft clears the stored set (explicit empty write, not silently kept).
- A note with only coded terms (no prose) still persists on Save & Next.
- Malformed/hostile `diagnoses` payloads are sanitized on read (unknown keys ignored, prototype-pollution-safe, per-section dedup + cap).

## Test plan

Unit + RTL: picker (debounce, stale-response guard, dedup/disabled rows, synonym hint, error path), sanitizer edge cases, save payload extension semantics (`{}` clearing), rehydration, chips-count-as-content, history rows — all mutation-checked. Post-merge workflow QA on dev.yosemitecrew.com.

## Alternatives considered

Inline rich-text mentions (heavier editor plugin; staged as follow-up). A separate diagnoses side panel (loses per-section provenance).

## Dependencies & risks

Depends on the deployed terms/suggest endpoint (live). Risk: suggest endpoint currently caps at 50 ranked matches server-side — acceptable for autocomplete.

### PR DESCRIPTION
**Pull request by ankit-yc**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/YosemiteCrew/Yosemite-Crew/blob/main/CONTRIBUTING.md#commit-message-convention.
- [x] There is an issue for the bug/feature this PR is for.
- [x] All existing tests and lints pass.

## What is the current behavior?

SOAP notes are free-text only: the four native S/O/A/P sections are rich-text editors with no connection to the clinical vocabulary. On dev, 62 SOAP notes exist and none carries a coded reference; the terms/suggest endpoint has no UI consumer; the SoapNote `diagnoses` JSON channel round-trips through the FHIR mapper but nothing ever writes it.

## What is the new behavior?

SOAP notes gain **coded-term autocomplete**: each native S/O/A/P section in the appointment workspace now carries a "Add coded term" picker that searches the Yosemite clinical vocabulary (11,742 terms, display + multilingual synonyms, ranked) via `GET /v1/codes/terms/suggest`, and pins picked concepts as removable chips. Picked terms persist through the SOAP note's existing `diagnoses` JSON channel (the `soap-note-diagnoses` composition extension) and rehydrate on reload; signed notes list their coded terms per section in the "All SOAP notes" history read-out.

- **`packages/types`**: `SoapCodedTerm` / `SoapCodedProblems` / `SOAP_CODED_SECTIONS` shared shapes plus `parseSoapCodedProblems` — a sanitizer that reads only the four known section keys (prototype-pollution-safe by construction), drops entries without code/label, trims, dedups per section, and caps at 50 terms/section. `SOAP_DIAGNOSES_EXTENSION_URL` is now exported so the frontend save path uses the mapper's exact constant.
- **`clinicalTermsService`** (new): typed wrapper for the terms/suggest endpoint.
- **`SoapCodedTermPicker`** (new): house search idiom (Search + SearchResultsDropdown + WorkspaceSearchResultRow), 250 ms debounce, stale-response guard, already-added rows disabled, synonym-match hint ("matches “Vómitos”") when the hit came from a translation.
- **`NativeSoapFields`**: one picker per section, domain-scoped — Subjective→PresentingComplaint, Assessment→Diagnosis, Plan→Procedure, Objective→all domains.
- **`SoapStep`**: chips ride the active draft via `upsertSoap`; history rows render "Coded terms (Assessment): Gastritis (YC-000123)".
- **`soapStepUtils`**: chips alone count as content, so a chips-only note persists instead of silently advancing unsaved.
- **`workspaceClinicalService`**: save always sends the diagnoses extension (explicit `{}` when empty so clearing chips on a draft PATCH actually clears the stored set); rehydration parses it back through the shared sanitizer.

No backend changes: the `diagnoses` channel already round-trips end-to-end (`compositionToSoapNoteInput` → `SoapNote.diagnoses` → `soapNoteToComposition`), so FHIR consumers of the SOAP Composition see the coded terms in the `soap-note-diagnoses` extension as of this PR.

## Why

The terminology programme (W1–W10) shipped the vocabulary, multilingual search, hierarchy, and cross-vocabulary mappings — but SOAP notes were still free-text only: 62 notes on dev, zero with any coded reference, and the suggest endpoint had no UI consumer. This is the last mile that turns clinician prose into coded, analysable, exportable records.

## Impact area

Appointment workspace SOAP step (web PIMS). Custom-template SOAP notes are unchanged (pickers render only for the native four-section layout). Mobile untouched.

## Validation performed

- `tsc --noEmit` clean (frontend, backend); `@yosemite-crew/types` builds.
- `pnpm --filter frontend run lint` clean.
- Targeted jest: SoapStep (28), workspaceClinicalService (24+2), SoapCodedTermPicker (10), clinicalTermsService (3), parseSoapCodedProblems (9) — all green.
- New-file coverage: SoapCodedTermPicker 100/100/100/100; clinicalTermsService 100%.
- **Mutation-checked (8 mutations, each makes exactly its pinning test fail):** parser hasOwnProperty guard / trim / dedup; save `?? {}` clearing semantics; rehydration parse; picker stale-response guard; chips-count-as-content; history read rows.
- Aikido opengrep SAST (309 rules) on all 8 changed source files: 0 findings. gitleaks on the full changed set: no leaks.
- Post-merge: full visual/workflow QA on dev.yosemitecrew.com (picker → chips → save → reload → history), screenshots to follow in a comment.

## Out of scope (follow-ups)

- Species filtering (vocabulary uses SA/LA/EQUINE… practice buckets; workspace has no companion→bucket mapping yet).
- Coded terms in the merged clinical-packet PDF and discharge summary.
- Inline in-text mentions (@-style) in the rich-text editors — chip rows are v1.
- Mobile SOAP surfaces.

## Related Issue(s)

Closes #2487

### CODE REVIEW

**Review by aupyay — DISMISSED** [review]

File: packages/types/src/clinical-artifact.ts [hunk 1]
```diff
@@ -45,6 +45,76 @@ export type SoapNoteInput = ClinicalArtifactBaseInput & {
...
+  for (const section of SOAP_CODED_SECTIONS) {
+    if (!Object.prototype.hasOwnProperty.call(source, section)) continue;
+    const raw = source[section];
+    if (!Array.isArray(raw)) continue;
+    const seen = new Set<string>();
+    const terms: SoapCodedTerm[] = [];
+    for (const item of raw) {
+      if (terms.length >= MAX_CODED_TERMS_PER_SECTION) break;
+      const term = parseSoapCodedTerm(item);
+      if (!term || seen.has(term.ycCode)) continue;
```
[the remark this answers is not in the document]
Comment (ankit-yc) [reply packages/types/src/clinical-artifact.ts]: Addressed in the follow-up commit: the per-section loop (array check, cap, dedup, term validation) is extracted into a parseSectionTerms helper, so parseSoapCodedProblems itself is now a flat iterate-sections-and-collect loop. Behaviour is unchanged and pinned by the 12 parser tests (including the prototype-pollution, dedup, and cap cases), which pass on the new shape.

**Review by aupyay — APPROVED** [review]
